design-documents/web-api-controllers.md

```diff
@@ -0,0 +1,289 @@
+{
+    //Now we have getId() since it's OK for storefront clients to know their customer's ID.
+    public function getId(): string;
+    
+    ....
+    //No getPassword()
+    ....
+}
+```
+ 
+By having these operation-specific DTOs we would only have the data we need as arguments and responses.
+ 
+###### How to introduce it to Magento?
+There are couple of ways we can introduce this concept to Magento:
+* use it for new storefront API
+* introduce new APIs to a module (for in stance _Customer_) in order to provide an example for 3rd-party developers
+* create a devdocs page/update technical guidelines to inform community
+ 
+ 
+#### Introduce reusable user input processors
+Sometimes before passing arguments to a service contracts we would have to perform user input processing that cannot
+be covered with declarative validation or existing ACL policies regarding of our current area (HTML/REST/SOAP/GraphQL).
+ 
+Example is checking whether is a customer allowed to use a saved shipping address for an order by checking whether the
+address was created by the same customer - we'd have to do it programmatically since validation rules wouldn't work
+because of not having access to the full context and our ACL system is not fit to process those cases. HTML controllers
+and GraphQL resolvers would be able to include such logic but it would not be reusable by other areas. This validation
+could be performed by the service contract responsible for posting order but service contracts are not supposed to know
+about authenticated users.
+ 
+The solution to these problems would be to introduce reusable area-agnostic user input processors that would perform
+such actions. These would be arbitrary classes with no common interface but they would be reusable across different areas
+and reside inside a special namespace `Presentation` to inform developers that this classes belong to presentation
+layer and are not supposed to be used inside other layers. These classes could be used explicitly by HTML controllers
+and graphQL resolvers and declared for REST/SOAP inside webapi.xml.
+ 
+ 
+In addition to the example above let's see how we could perform the ownership validation for shipping address with
+a user input processor.
+ 
+Let's say we have a service contracts responsible for updating shipping address for an order:
+```php
+interface OrderManagerInterface
+{
+    ....
+    
+    public function updateShippingAddress(string $cartId, CustomerAddressInterface $shippingAddress): void
+}
+```
+ 
+and then we create user input processor to check the ownership:
+```php
+namespace Magento\Sales\Presentation;
+
+class OrderShippingAddressValidator
+{
+    /**
+     * @var UserContextInterface
+     */
+    private $userContext;
+    
+    .....
+
+    /**
+     * @throws SecurityViolationException
+     */
+    public function validateAddressForOrder(CustomerAddressInterface $shippingAddress): void
+    {
+        $customerId = $this->userContext->getUserId();
```

Comment:
How the validator will receive the `$userContext` if it injects via constructor? For example, GraphQl resolver receives it via `resolve` method and the validator instance already should be initialized.

Reply:
the default implementation of UserContextInterface is a composite of all possible context classes which cycles them until one of them can return a user, I assume for GraphQL context it will use CustomerSessionUserContext to get the user on runtime when validation is invoked

Reply:
My question is how the context dependency will be initialized because when GraphQl resolver calls `$this->addressValidator->validateAddressForOrder()` in `resolve` method the `OrderShippingAddressValidator` object should be already initialized with the `Context` dependency. So, probably, `OrderShippingAddressValidator` should receive context via method argument instead of constructor?